sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -359,24 +364,44 @@ def test_glm_regression_unpenalized(solver, fit_intercept, glm_dataset):
     else:
         intercept = 0
 
-    model.fit(X, y)
+    with warnings.catch_warnings():
+        if solver.startswith("newton") and n_samples < n_features:
+            # The newton solvers should warn and automatically fallback to LBFGS
+            # in this case. The model should still converge.
+            warnings.filterwarnings("ignore", category=scipy.linalg.LinAlgWarning)
+        # XXX: Investigate if the ConvergenceWarning that can appear in some
+        # cases should be considered a bug or not. In the mean time we don't
+        # fail when the assertions below pass irrespective of the presence of
+        # the warning.
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
+        model.fit(X, y)
 
     # FIXME: `assert_allclose(model.coef_, coef)` should work for all cases but fails
     # for the wide/fat case with n_features > n_samples. Most current GLM solvers do
     # NOT return the minimum norm solution with fit_intercept=True.
-    rtol = 5e-5
     if n_samples > n_features:
+        rtol = 5e-5 if solver == "lbfgs" else 1e-7
         assert model.intercept_ == pytest.approx(intercept)
         assert_allclose(model.coef_, coef, rtol=rtol)
     else:
         # As it is an underdetermined problem, prediction = y. The following shows that
         # we get a solution, i.e. a (non-unique) minimum of the objective function ...
-        assert_allclose(model.predict(X), y, rtol=1e-6)
-        if fit_intercept:
+        rtol = 5e-5
+        if solver == "newton-cholesky":
+            rtol = 5e-4
+        assert_allclose(model.predict(X), y, rtol=rtol)
+
+        norm_solution = np.linalg.norm(np.r_[intercept, coef])
+        norm_model = np.linalg.norm(np.r_[model.intercept_, model.coef_])
+        if solver == "newton-cholesky":
+            # XXX: This solver shows random behaviour. Sometimes it finds solutions
+            # with norm_model <= norm_solution! So we check conditionally.
+            if not (norm_model > (1 + 1e-12) * norm_solution):
```

Comment:
The following looks more intuitive.

```suggestion
            if norm_model < (1 + 1e-12) * norm_solution:
```

But then it means that we make no assertion on attempting to reach the minimum norm solution. It could be the case that this model never reaches the minimum norm solution (in particular when `fit_intercept is True`) and we will likely face a similar problem as #23670 when using the smart (non-zero) intercept initialization strategy.

Reply:
As written in the comment: Sometimes it finds the minimum norm solution (strangely in this case, the norm is sometimes even smaller than the minimum norm! I suspect this to be caused by numerical precision). Sometimes it does not find the minimum norm solution (which happens more often).